jena-arq/src/main/java/org/apache/jena/sparql/core/assembler/InMemDatasetAssembler.java

```diff
@@ -0,0 +1,67 @@
+package org.apache.jena.sparql.core.assembler;
+
+import static org.apache.jena.assembler.JA.MemoryDataset;
+import static org.apache.jena.assembler.JA.data;
+import static org.apache.jena.query.DatasetFactory.createTxnMem;
+import static org.apache.jena.riot.RDFDataMgr.loadModel;
+import static org.apache.jena.riot.RDFDataMgr.read;
+import static org.apache.jena.sparql.core.assembler.AssemblerUtils.setContext;
+import static org.apache.jena.sparql.core.assembler.DatasetAssemblerVocab.*;
+import static org.apache.jena.sparql.util.FmtUtils.stringForRDFNode;
+import static org.apache.jena.sparql.util.graph.GraphUtils.*;
+
+import java.util.function.Predicate;
+
+import org.apache.jena.assembler.Assembler;
+import org.apache.jena.assembler.Mode;
+import org.apache.jena.assembler.assemblers.AssemblerBase;
+import org.apache.jena.query.Dataset;
+import org.apache.jena.rdf.model.Model;
+import org.apache.jena.rdf.model.RDFNode;
+import org.apache.jena.rdf.model.Resource;
+
+/**
+ * An {@link Assembler} that creates in-memory {@link Dataset}s.
+ *
+ */
+public class InMemDatasetAssembler extends AssemblerBase {
+
+	@Override
+	public Dataset open(final Assembler assembler, final Resource root, final Mode mode) {
+		checkType(root, MemoryDataset);
+		final Dataset dataset = createTxnMem();
+		setContext(root, dataset.getContext());
+		// Default graph can be defined with ja:graph or ja:defaultGraph
+		final Resource defaultGraphDef = root.hasProperty(pDefaultGraph) ? getResourceValue(root,
+				pDefaultGraph) : root.hasProperty(pGraph) ? getResourceValue(root, pGraph) : null;
```

Comment:
Okay, I thought it would be cool, but I certainly admit that it's gotten confusing already, and no users have even gotten involved yet! {grin}

I will strip out this jazz entirely and take it back to the plain case. To do that, I will use only one predicate, `ja:data`, with the meaning of "load tuples from this URI".